Error on upgrade after fresh installation

### Expected behavior

As a IDEasy user, I want to call `ide upgrade` so that I can check for the latest version and get it installed if a newer release is available.

### Actual behavior

```
$ ide upgrade
Updates are available for the settings repository. If you want to apply the latest changes, call "ide update"
Start: Install ideasy
Deleting corrupted installation at C:\Users\joerg\projects\_ide\software\maven\ideasy\ideasy\2025.03.002
Step 'ide (upgrade)' failed: java.lang.IllegalStateException: Failed to delete C:\Users\joerg\projects\_ide\software\maven\ideasy\ideasy\2025.03.002
 Step 'Install ideasy (*)' failed: java.lang.IllegalStateException: Failed to delete C:\Users\joerg\projects\_ide\software\maven\ideasy\ideasy\2025.03.002
2 step(s) failed out of 2 steps.
An unexpected error occurred!
We are sorry for the inconvenience.
Please check the error below, resolve it and try again.
If the error is not on your end (network connectivity, lack of permissions, etc.) please file a bug:
https://github.com/devonfw/IDEasy/issues/new?assignees=&labels=bug&projects=&template=bug.md&title=IllegalStateException%3A+Failed+to+delete+C%3A%5CUsers%5Cjoerg%5Cprojects%5C_ide%5Csoftware%5Cmaven%5Cideasy%5Cideasy%5C2025.03.002
java.lang.IllegalStateException: Failed to delete C:\Users\joerg\projects\_ide\software\maven\ideasy\ideasy\2025.03.002
        at com.devonfw.tools.ide.io.FileAccessImpl.delete(FileAccessImpl.java:791)
        at com.devonfw.tools.ide.tool.LocalToolCommandlet.installTool(LocalToolCommandlet.java:208)
        at com.devonfw.tools.ide.tool.LocalToolCommandlet.installTool(LocalToolCommandlet.java:169)
        at com.devonfw.tools.ide.tool.LocalToolCommandlet.install(LocalToolCommandlet.java:81)
        at com.devonfw.tools.ide.tool.ToolCommandlet.install(ToolCommandlet.java:260)
        at com.devonfw.tools.ide.tool.IdeasyCommandlet.install(IdeasyCommandlet.java:109)
        at com.devonfw.tools.ide.commandlet.UpgradeCommandlet.run(UpgradeCommandlet.java:43)
        at com.devonfw.tools.ide.context.AbstractIdeContext.applyAndRun(AbstractIdeContext.java:958)
        at com.devonfw.tools.ide.context.AbstractIdeContext.run(AbstractIdeContext.java:876)
        at com.devonfw.tools.ide.cli.Ideasy.runOrThrow(Ideasy.java:95)
        at com.devonfw.tools.ide.cli.Ideasy.run(Ideasy.java:59)
        at com.devonfw.tools.ide.cli.Ideasy.main(Ideasy.java:133)
        at java.base@21.0.2/java.lang.invoke.LambdaForm$DMH/sa346b79c.invokeStaticInit(LambdaForm$DMH)
Caused by: java.nio.file.AccessDeniedException: C:\Users\joerg\projects\_ide\software\maven\ideasy\ideasy\2025.03.002\bin\ideasy.exe
        at java.base@21.0.2/sun.nio.fs.WindowsFileSystemProvider.implDelete(WindowsFileSystemProvider.java:273)
        at java.base@21.0.2/sun.nio.fs.AbstractFileSystemProvider.delete(AbstractFileSystemProvider.java:104)
        at java.base@21.0.2/java.nio.file.Files.delete(Files.java:1152)
        at com.devonfw.tools.ide.io.FileAccessImpl.deleteRecursive(FileAccessImpl.java:807)
        at com.devonfw.tools.ide.io.FileAccessImpl.deleteRecursive(FileAccessImpl.java:802)
        at com.devonfw.tools.ide.io.FileAccessImpl.deleteRecursive(FileAccessImpl.java:802)
        at com.devonfw.tools.ide.io.FileAccessImpl.delete(FileAccessImpl.java:788)
        ... 12 more

Error: IDEasy failed with exit code 255
```

Why does IDEasy consider my fresh installation created via MSI installer as corrupted?

### Steps to reproduce (bug) / Use Case of feature request (enhancement)

1. I had an earlier SNAPSHOT installed on my ARM based windows machine.
2. I wanted to upgrade but that old SNAPSHOT version did not support `ide upgrade`. Therefore I just downloaded the latest MSI and ran the installer.
3. I opened a console and tried `ide upgrade` (for curiosity)

### Related/Dependent Issues


### Comments/Hints:

If I just created the current installation via MSI installer it should not be considered as "corrupted".
This is one of the few rare cases, where Windows file locks are actually of any good since they prevented IDEasy from deleting itself.
Why after all does IDEasy delete itself even before installing a new version of IDEasy?
This sounds like the perfect plan for disaster (on Linux or Mac now IDEasy would have deleted itself and instead of upgrading I would have uninstalled IDEasy).

### Affected version:

* Tested on Windows, maybe all OS affected
* 2025.03.002

--- a/cli/src/main/java/com/devonfw/tools/ide/context/AbstractIdeContext.java
+++ b/cli/src/main/java/com/devonfw/tools/ide/context/AbstractIdeContext.java
@@ -1317,4 +1317,13 @@ public abstract class AbstractIdeContext implements IdeContext {
     this.variables = null;
     this.customToolRepository = null;
   }
+
+  @Override
+  public void writeVersionFile(VersionIdentifier version, Path installationPath) {
+
+    assert (Files.isDirectory(installationPath));
+    Path versionFile = installationPath.resolve(FILE_SOFTWARE_VERSION);
+    getFileAccess().writeFileContent(version.toString(), versionFile);
+  }
+
 }

--- a/cli/src/main/java/com/devonfw/tools/ide/context/IdeContext.java
+++ b/cli/src/main/java/com/devonfw/tools/ide/context/IdeContext.java
@@ -719,4 +719,10 @@ public interface IdeContext extends IdeStartContext {
    */
   void logIdeHomeAndRootStatus();
 
+  /**
+   * @param version the {@link VersionIdentifier} to write.
+   * @param installationPath the {@link Path directory} where to write the version to a {@link #FILE_SOFTWARE_VERSION version file}.
+   */
+  void writeVersionFile(VersionIdentifier version, Path installationPath);
+
 }

--- a/cli/src/main/java/com/devonfw/tools/ide/tool/IdeasyCommandlet.java
+++ b/cli/src/main/java/com/devonfw/tools/ide/tool/IdeasyCommandlet.java
@@ -173,6 +173,7 @@ public class IdeasyCommandlet extends MvnBasedLocalToolCommandlet {
     for (Path installationArtifact : installationArtifacts) {
       fileAccess.copy(installationArtifact, ideasyVersionPath);
     }
+    this.context.writeVersionFile(IdeVersion.getVersionIdentifier(), ideasyVersionPath);
     fileAccess.symlink(ideasyVersionPath, installationPath);
     addToShellRc(BASHRC, ideRoot, null);
     addToShellRc(ZSHRC, ideRoot, "autoload -U +X bashcompinit && bashcompinit");

--- a/cli/src/main/java/com/devonfw/tools/ide/tool/LocalToolCommandlet.java
+++ b/cli/src/main/java/com/devonfw/tools/ide/tool/LocalToolCommandlet.java
@@ -1,10 +1,8 @@
 package com.devonfw.tools.ide.tool;
 
-import java.io.IOException;
 import java.nio.file.Files;
 import java.nio.file.LinkOption;
 import java.nio.file.Path;
-import java.nio.file.StandardOpenOption;
 import java.util.Collection;
 import java.util.Set;
 
@@ -223,11 +221,7 @@ public abstract class LocalToolCommandlet extends ToolCommandlet {
     }
     fileAccess.mkdirs(installationPath.getParent());
     fileAccess.extract(downloadedToolFile, installationPath, this::postExtract, extract);
-    try {
-      Files.writeString(toolVersionFile, resolvedVersion.toString(), StandardOpenOption.CREATE_NEW);
-    } catch (IOException e) {
-      throw new IllegalStateException("Failed to write version file " + toolVersionFile, e);
-    }
+    this.context.writeVersionFile(resolvedVersion, installationPath);
     this.context.debug("Installed {} in version {} at {}", this.tool, resolvedVersion, installationPath);
     return createToolInstallation(installationPath, resolvedVersion, toolVersionFile, true, processContext, extraInstallation);
   }

--- a/cli/src/test/java/com/devonfw/tools/ide/tool/IdeasyCommandletTest.java
+++ b/cli/src/test/java/com/devonfw/tools/ide/tool/IdeasyCommandletTest.java
@@ -7,6 +7,7 @@ import org.junit.jupiter.params.ParameterizedTest;
 import org.junit.jupiter.params.provider.ValueSource;
 
 import com.devonfw.tools.ide.context.AbstractIdeContextTest;
+import com.devonfw.tools.ide.context.IdeContext;
 import com.devonfw.tools.ide.context.IdeTestContext;
 import com.devonfw.tools.ide.os.SystemInfo;
 import com.devonfw.tools.ide.os.SystemInfoMock;
@@ -89,6 +90,7 @@ public class IdeasyCommandletTest extends AbstractIdeContextTest {
     assertThat(installationPath.resolve("IDEasy.pdf")).hasContent("IDEasy.pdf mock");
     assertThat(installationPath.resolve("setup")).hasContent("setup mock");
     assertThat(installationPath.resolve("setup.bat")).hasContent("setup.bat mock");
+    assertThat(installationPath.resolve(IdeContext.FILE_SOFTWARE_VERSION)).hasContent("SNAPSHOT");
   }
 
 }

